src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceSetDeserializer.cs

```diff
@@ -147,9 +167,143 @@ public virtual IEnumerable ReadResourceSet(ODataResourceSetWrapper resourceSet,
                     Error.Format(SRResources.TypeCannotBeDeserialized, elementType.FullName()));
             }
 
-            foreach (ODataResourceWrapper resourceWrapper in resourceSet.Resources)
+            //Ideally we don't need to special case ResourceSetType.ResourceSet, since the code that handles a deltaresourcesetwrapper will also handle a ResourceSetWrapper, 
+            //but it may be more efficient for the common case.
+
+            if (resourceSet.ResourceSetType == ResourceSetType.ResourceSet)
+            {
+                foreach (ODataResourceWrapper resourceWrapper in resourceSet.Items)
+                {
+                    yield return deserializer.ReadInline(resourceWrapper, elementType, readContext);
+                }
+            }
+            else
+            {
+                Type clrType = EdmLibHelpers.GetClrType(elementType, readContext.Model);
+
+                if (readContext.IsChangedObjectCollection)
+                {
+                    readContext.ResourceType = typeof(Delta<>).MakeGenericType(clrType);
+                }
+
+                foreach (ODataItemBase odataItemBase in resourceSet.Items)
+                {
+                    ODataResourceWrapper resourceWrapper = odataItemBase as ODataResourceWrapper;
+
+                    if(resourceWrapper != null)
+                    {
+                        ODataDeletedResource deletedResource = resourceWrapper.ResourceBase as ODataDeletedResource;
+
+                        if (deletedResource != null)
+                        {
+                            yield return DeSerializeDeletedEntity(deletedResource, readContext, clrType);
```

Comment:
>DeSerializeDeletedEntity [](start = 41, length = 24)

Why did we put the logic for deserializing a deleted entity in the ODataResourceSetDeserializer, rather than handling deleted resources in the ODataResourceSerializer?  Then we could get rid of the special handling here.